crypto/dilithium/pqcrystals_dilithium_ref_common/rounding.c

```diff
@@ -31,24 +31,27 @@ int32_t power2round(int32_t *a0, int32_t a)  {
 *              -ALPHA/2 <= a0 = a mod^+ Q - Q < 0. Assumes a to be standard
 *              representative.
 *
-* Arguments:   - int32_t a: input element
+* Arguments:   - ml_dsa_params: parameter struct
+*              - int32_t a: input element
 *              - int32_t *a0: pointer to output element a0
 *
 * Returns a1.
 **************************************************/
-int32_t decompose(int32_t *a0, int32_t a) {
+int32_t decompose(ml_dsa_params *params, int32_t *a0, int32_t a) {
+  assert((params->gamma2 == (Q-1)/32) || (params->gamma2 == (Q-1)/88));
+
   int32_t a1;
 
   a1  = (a + 127) >> 7;
-#if GAMMA2 == (Q-1)/32
-  a1  = (a1*1025 + (1 << 21)) >> 22;
-  a1 &= 15;
-#elif GAMMA2 == (Q-1)/88
-  a1  = (a1*11275 + (1 << 23)) >> 24;
-  a1 ^= ((43 - a1) >> 31) & a1;
-#endif
-
-  *a0  = a - a1*2*GAMMA2;
+  if (params->gamma2 == (Q-1)/32) {
+    a1  = (a1*1025 + (1 << 21)) >> 22;
+    a1 &= 15;
+  }
+  if (params->gamma2 == (Q-1)/88) {
+    a1  = (a1*11275 + (1 << 23)) >> 24;
+    a1 ^= ((43 - a1) >> 31) & a1;
+  }
```

Comment:
Is gamm2 guaranteed to always be one or the other value? I'm thinking about ways to simplify this and make it more clear.

Reply:
It is, we assert so on line 41 above. `gamma2` is defined on page 15 of https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.204.pdf and can only be `(Q-1)/88` or `(Q-1)/32`.

One option is to add:
```
if (gamma2 == (Q-1)/32) {...}
else if (gamma2 == (Q-1)/88) {...}
else {return fail}
```
However, `decompose` does not return a value, and subsequent calling functions up the chain do not have the ability to handle failures.